File: backend/src/modules/openai-compatible-files/files.routes.ts
import { createHash } from 'node:crypto'
import { createReadStream, createWriteStream } from 'node:fs'
import { pipeline } from 'node:stream/promises'
import { Transform } from 'node:stream'

import Busboy from 'busboy'
import { Router, type NextFunction, type Request, type Response } from 'express'

import { requestDbService } from '../db-service/db-service-ipc.js'
import type { GatewayRuntimeRequest } from '../gateway/request/pre-auth.js'
import { gatewayErrorPayload } from '../gateway/response/responses.js'
import type { OpenAICompatibleFileRecord } from '../../storage/openai-compatible-files.repository.js'
import {
  ensureOpenAICompatibleFileObjectParent,
  newOpenAICompatibleFileId,
  normalizeOpenAICompatibleFileMediaType,
  openAICompatibleFileMaxBytes,
  openAICompatibleFileObjectPath,
  removeOpenAICompatibleFileObject,
  storageKeyForOpenAICompatibleFile
} from './file-storage.js'

export const openAICompatibleFilesRouter = Router()

interface UploadedOpenAICompatibleFile {
  fileId: string
  filename: string
  mediaType?: string
  storageKey: string
  bytes: number
  sha256: string
}

class OpenAICompatibleFilesRequestError extends Error {
  constructor(
    message: string,
    readonly statusCode = 400,
    readonly type = 'invalid_request_error',
    readonly code?: string
  ) {
    super(message)
  }
}

openAICompatibleFilesRouter.get('/v1/files', handleOpenAICompatibleFilesRoute(listOpenAICompatibleFiles))
openAICompatibleFilesRouter.post('/v1/files', handleOpenAICompatibleFilesRoute(uploadOpenAICompatibleFile))
openAICompatibleFilesRouter.get('/v1/files/:fileId/content', handleOpenAICompatibleFilesRoute(downloadOpenAICompatibleFileContent))
openAICompatibleFilesRouter.get('/v1/files/:fileId', handleOpenAICompatibleFilesRoute(getOpenAICompatibleFile))
openAICompatibleFilesRouter.delete('/v1/files/:fileId', handleOpenAICompatibleFilesRoute(deleteOpenAICompatibleFileRoute))

async function uploadOpenAICompatibleFile(req: Request, res: Response): Promise<void> {
  const runtime = requireGatewayRuntime(req)
  const contentType = req.header('content-type') ?? ''
  if (!/^multipart\/form-data\b/i.test(contentType)) {
    throw new OpenAICompatibleFilesRequestError('Files upload requires multipart/form-data', 400, 'invalid_request_error', 'invalid_content_type')
  }
  const upload = await readOpenAICompatibleMultipartUpload(req)
  let created: OpenAICompatibleFileRecord | undefined
  try {
    created = await requestDbService({
      type: 'create_openai_compatible_file',
      input: {
        id: upload.file.fileId,
        systemAccountId: runtime.apiKey.system_account_id,
        apiKeyId: runtime.apiKey.id,
        purpose: upload.purpose,
        filename: upload.file.filename,
        bytes: upload.file.bytes,
        mediaType: upload.file.mediaType,
        storageKey: upload.file.storageKey,
        sha256: upload.file.sha256
      }
    }, { timeoutMs: 10_000 })
  } catch (error) {
    await removeOpenAICompatibleFileObject(upload.file.storageKey).catch(() => undefined)
    throw error
  }
  res.status(200).json(openAICompatibleFileObject(created))
}

async function listOpenAICompatibleFiles(req: Request, res: Response): Promise<void> {
  const runtime = requireGatewayRuntime(req)
  const result = await requestDbService({
    type: 'list_openai_compatible_files',
    options: {
      systemAccountId: runtime.apiKey.system_account_id,
      apiKeyId: runtime.apiKey.id,
      purpose: queryString(req.query.purpose),
      limit: queryInteger(req.query.limit),
      order: queryString(req.query.order) === 'asc' ? 'asc' : 'desc',
      after: queryString(req.query.after)
    }
  })
  const data = result.items.map(openAICompatibleFileObject)
  res.json({
    object: 'list',
    data,
    first_id: data[0]?.id,
    last_id: data[data.length - 1]?.id,
    has_more: result.hasMore
  })
}

async function getOpenAICompatibleFile(req: Request, res: Response): Promise<void> {
  const record = await findOpenAICompatibleFileForRequest(req)
  if (!record) {
    throw new OpenAICompatibleFilesRequestError('File not found', 404, 'invalid_request_error', 'file_not_found')
  }
  res.json(openAICompatibleFileObject(record))
}

async function downloadOpenAICompatibleFileContent(req: Request, res: Response): Promise<void> {
  const record = await findOpenAICompatibleFileForRequest(req)
  if (!record) {
    throw new OpenAICompatibleFilesRequestError('File not found', 404, 'invalid_request_error', 'file_not_found')
  }
  const filePath = openAICompatibleFileObjectPath(record.storageKey)
  res.status(200)
  res.setHeader('content-type', record.mediaType ?? 'application/octet-stream')
  res.setHeader('content-length', String(record.bytes))
  await pipeline(createReadStream(filePath), res)
}

async function deleteOpenAICompatibleFileRoute(req: Request, res: Response): Promise<void> {
  const runtime = requireGatewayRuntime(req)
  const fileId = pathFileId(req)
  const record = await requestDbService({
    type: 'delete_openai_compatible_file',
    fileId,
    systemAccountId: runtime.apiKey.system_account_id,
    apiKeyId: runtime.apiKey.id
  })
  if (!record) {
    throw new OpenAICompatibleFilesRequestError('File not found', 404, 'invalid_request_error', 'file_not_found')
  }
  await removeOpenAICompatibleFileObject(record.storageKey).catch(() => undefined)
  res.json({
    id: record.id,
    object: 'file',
    deleted: true
  })
}

async function findOpenAICompatibleFileForRequest(req: Request): Promise<OpenAICompatibleFileRecord | undefined> {
  const runtime = requireGatewayRuntime(req)
  return await requestDbService({
    type: 'get_openai_compatible_file',
    fileId: pathFileId(req),
    systemAccountId: runtime.apiKey.system_account_id,
    apiKeyId: runtime.apiKey.id
  })
}

async function readOpenAICompatibleMultipartUpload(req: Request): Promise<{
  purpose: string
  file: UploadedOpenAICompatibleFile
}> {
  let purpose = ''
  const fileUploads: Array<Promise<UploadedOpenAICompatibleFile>> = []
  let parseError: OpenAICompatibleFilesRequestError | undefined
  const busboy = Busboy({
    headers: req.headers,
    limits: {
      files: 1,
      fields: 8,
      fileSize: openAICompatibleFileMaxBytes
    }
  })
  busboy.on('field', (name, value) => {
    if (name === 'purpose') {
      purpose = String(value ?? '').trim()
    }
  })
  busboy.on('file', (name, stream, info) => {
    if (name !== 'file') {
      stream.resume()
      return
    }
    if (fileUploads.length > 0) {
      parseError = new OpenAICompatibleFilesRequestError('Only one file can be uploaded per request', 400, 'invalid_request_error', 'too_many_files')
      stream.resume()
      return
    }
    fileUploads.push(writeUploadedOpenAICompatibleFile(stream, info))
  })
  busboy.on('filesLimit', () => {
    parseError = new OpenAICompatibleFilesRequestError('Only one file can be uploaded per request', 400, 'invalid_request_error', 'too_many_files')
  })
  await new Promise<void>((resolve, reject) => {
    busboy.once('finish', resolve)
    busboy.once('error', reject)
    req.pipe(busboy)
  })
  if (parseError) throw parseError
  const files = await Promise.all(fileUploads)
  const file = files[0]
  if (!purpose) {
    if (file) await removeOpenAICompatibleFileObject(file.storageKey).catch(() => undefined)
    throw new OpenAICompatibleFilesRequestError('Missing required multipart field: purpose', 400, 'invalid_request_error', 'missing_purpose')
  }
  if (!file) {
    throw new OpenAICompatibleFilesRequestError('Missing required multipart file field: file', 400, 'invalid_request_error', 'missing_file')
  }
  return { purpose, file }
}

async function writeUploadedOpenAICompatibleFile(
  stream: NodeJS.ReadableStream,
  info: { filename?: string; mimeType?: string }
): Promise<UploadedOpenAICompatibleFile> {
  const fileId = newOpenAICompatibleFileId()
  const filename = normalizedUploadFilename(info.filename)
  const mediaType = normalizeOpenAICompatibleFileMediaType(info.mimeType, filename)
  const storageKey = storageKeyForOpenAICompatibleFile(fileId)
  const filePath = ensureOpenAICompatibleFileObjectParent(storageKey)
  const hash = createHash('sha256')
  let bytes = 0
  let fileSizeLimitHit = false
  stream.on('limit', () => {
    fileSizeLimitHit = true
  })
  const counter = new Transform({
    transform(chunk: Buffer, _encoding, callback) {
      bytes += chunk.length
      hash.update(chunk)
      callback(null, chunk)
    }
  })
  try {
    await pipeline(stream, counter, createWriteStream(filePath))
    if (fileSizeLimitHit) {
      throw new OpenAICompatibleFilesRequestError('Uploaded file is too large', 413, 'request_too_large', 'file_too_large')
    }
    if (bytes <= 0) {
      throw new OpenAICompatibleFilesRequestError('Uploaded file is empty', 400, 'invalid_request_error', 'empty_file')
    }
    return {
      fileId,
      filename,
      mediaType,
      storageKey,
      bytes,
      sha256: hash.digest('hex')
    }
  } catch (error) {
    await removeOpenAICompatibleFileObject(storageKey).catch(() => undefined)
    throw error
  }
}

function handleOpenAICompatibleFilesRoute(
  handler: (req: Request, res: Response) => Promise<void>
): (req: Request, res: Response, next: NextFunction) => void {
  return (req, res, next) => {
    handler(req, res).catch((error: unknown) => {
      if (res.headersSent) {
        next(error)
        return
      }
      if (error instanceof OpenAICompatibleFilesRequestError) {
        res.status(error.statusCode).json(gatewayErrorPayload(error.message, error.type, error.code))
        return
      }
      next(error)
    })
  }
}

function requireGatewayRuntime(req: Request): Required<Pick<NonNullable<GatewayRuntimeRequest['gatewayRuntime']>, 'apiKey'>> {
  const runtime = (req as GatewayRuntimeRequest).gatewayRuntime
  if (!runtime?.apiKey) {
    throw new OpenAICompatibleFilesRequestError('Missing or invalid API key', 401, 'invalid_request_error', 'invalid_api_key')
  }
  return { apiKey: runtime.apiKey }
}

function openAICompatibleFileObject(record: OpenAICompatibleFileRecord): Record<string, unknown> {
  return {
    id: record.id,
    object: 'file',
    bytes: record.bytes,
    created_at: openAITimestamp(record.createdAt),
    filename: record.filename,
    purpose: record.purpose,
    status: record.status,
    ...(record.expiresAt ? { expires_at: openAITimestamp(record.expiresAt) } : {})
  }
}

function pathFileId(req: Request): string {
  const value = typeof req.params.fileId === 'string' ? req.params.fileId.trim() : ''
  if (!value) {
    throw new OpenAICompatibleFilesRequestError('Missing file id', 400, 'invalid_request_error', 'missing_file_id')
  }
  return value
}

function normalizedUploadFilename(value: string | undefined): string {
  const filename = value?.split(/[\\/]/).pop()?.trim()
  return filename || 'upload'
}

function queryString(value: unknown): string | undefined {
  if (typeof value === 'string') {
    const text = value.trim()
    return text || undefined
  }
  return undefined
}

function queryInteger(value: unknown): number | undefined {
  const text = queryString(value)
  if (!text) return undefined
  const number = Number(text)
  return Number.isFinite(number) ? Math.trunc(number) : undefined
}

function openAITimestamp(value: string): number {
  const ms = Date.parse(value)
  return Number.isFinite(ms) ? Math.floor(ms / 1000) : Math.floor(Date.now() / 1000)
}

File: backend-go/projects/gateway/internal/openaicompat/filesroutes.go
package openaicompat

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"io"
	"mime"
	"mime/multipart"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
)

// Files routes port openai-compatible-files/files.routes.ts byte-for-byte:
// envelope key order, 中文错误文案 and status codes follow the Node contract.

type fileObject struct {
	ID        string `json:"id"`
	Object    string `json:"object"`
	Bytes     int64  `json:"bytes"`
	CreatedAt int64  `json:"created_at"`
	Filename  string `json:"filename"`
	Purpose   string `json:"purpose"`
	Status    string `json:"status"`
	ExpiresAt *int64 `json:"expires_at,omitempty"`
}

type containerFileObject struct {
	ID          string  `json:"id"`
	Object      string  `json:"object"`
	CreatedAt   int64   `json:"created_at"`
	Bytes       int64   `json:"bytes"`
	Filename    string  `json:"filename"`
	ContainerID *string `json:"container_id"`
	Purpose     string  `json:"purpose"`
	Status      string  `json:"status"`
}

type listEnvelope struct {
	Object  string `json:"object"`
	Data    []any  `json:"data"`
	FirstID any    `json:"first_id,omitempty"`
	LastID  any    `json:"last_id,omitempty"`
	HasMore bool   `json:"has_more"`
}

type deletedFileEnvelope struct {
	ID      string `json:"id"`
	Object  string `json:"object"`
	Deleted bool   `json:"deleted"`
}

// MountFiles registers the /v1/files + /v1/containers file family.
func (d *Deps) MountFiles(k interface {
	Register(pattern string, handler http.Handler)
}) {
	k.Register("GET /v1/containers/{containerId}/files", d.guard(d.listContainerFiles))
	k.Register("GET /v1/containers/{containerId}/files/{fileId}/content", d.guard(d.downloadContainerFileContent))
	k.Register("GET /v1/containers/{containerId}/files/{fileId}", d.guard(d.getContainerFile))
	k.Register("GET /v1/files", d.guard(d.listFiles))
	k.Register("POST /v1/files", d.guard(d.uploadFile))
	k.Register("GET /v1/files/{fileId}/content", d.guard(d.downloadFileContent))
	k.Register("GET /v1/files/{fileId}", d.guard(d.getFile))
	k.Register("DELETE /v1/files/{fileId}", d.guard(d.deleteFile))
}

// guard wraps a handler with the shared error contract.
func (d *Deps) guard(run func(w http.ResponseWriter, r *http.Request) error) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		handle(w, func() error { return run(w, r) })
	})
}

// newFileObject mirrors openAICompatibleFileObject.
func newFileObject(record FileRecord) (fileObject, error) {
	createdAt, err := openAITimestamp(record.CreatedAt)
	if err != nil {
		return fileObject{}, err
	}
	object := fileObject{
		ID:        record.ID,
		Object:    "file",
		Bytes:     record.Bytes,
		CreatedAt: createdAt,
		Filename:  record.Filename,
		Purpose:   record.Purpose,
		Status:    record.Status,
	}
	if record.ExpiresAt != nil {
		expiresAt, err := openAITimestamp(*record.ExpiresAt)
		if err != nil {
			return fileObject{}, err
		}
		object.ExpiresAt = &expiresAt
	}
	return object, nil
}

// newContainerFileObject mirrors openAICompatibleContainerFileObject.
func newContainerFileObject(record FileRecord) (containerFileObject, error) {
	createdAt, err := openAITimestamp(record.CreatedAt)
	if err != nil {
		return containerFileObject{}, err
	}
	return containerFileObject{
		ID:          record.ID,
		Object:      "container.file",
		CreatedAt:   createdAt,
		Bytes:       record.Bytes,
		Filename:    record.Filename,
		ContainerID: record.ContainerID,
		Purpose:     record.Purpose,
		Status:      record.Status,
	}, nil
}

func writeJSONBody(w http.ResponseWriter, status int, payload any) {
	encoded, err := json.Marshal(payload)
	if err != nil {
		writeUnhandledError(w)
		return
	}
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(status)
	_, _ = w.Write(encoded)
}

func writeListEnvelope(w http.ResponseWriter, data []any) {
	envelope := listEnvelope{Object: "list", Data: data}
	if len(data) > 0 {
		if first, ok := data[0].(interface{ idValue() string }); ok {
			envelope.FirstID = first.idValue()
			envelope.LastID = data[len(data)-1].(interface{ idValue() string }).idValue()
		}
	}
	writeJSONBody(w, http.StatusOK, envelope)
}

func (o fileObject) idValue() string          { return o.ID }
func (o containerFileObject) idValue() string { return o.ID }

// pathFileID mirrors pathFileId: 缺少文件 ID -> 400 missing_file_id.
func pathFileID(r *http.Request) (string, error) {
	value := strings.TrimSpace(r.PathValue("fileId"))
	if value == "" {
		return "", badRequest("缺少文件 ID", "missing_file_id")
	}
	return value, nil
}

// pathContainerID mirrors pathContainerId: 缺少容器 ID -> 400.
func pathContainerID(r *http.Request) (string, error) {
	value := strings.TrimSpace(r.PathValue("containerId"))
	if value == "" {
		return "", badRequest("缺少容器 ID", "missing_container_id")
	}
	return value, nil
}

func (d *Deps) listFiles(w http.ResponseWriter, r *http.Request) error {
	scope := d.requireScope(w, r)
	if scope == nil {
		return nil
	}
	result, err := d.Store.ListFiles(r.Context(), FileListOptions{
		SystemAccountID: scope.SystemAccountID,
		APIKeyID:        scope.APIKeyID,
		Purpose:         queryStringParam(r.URL.Query(), "purpose"),
		Limit:           queryIntegerParam(r.URL.Query(), "limit"),
		Order:           textParam(r.URL.Query(), "order"),
		After:           textParamOrEmpty(r.URL.Query(), "after"),
	})
	if err != nil {
		return err
	}
	data := make([]any, 0, len(result.Items))
	for _, item := range result.Items {
		object, err := newFileObject(item)
		if err != nil {
			return err
		}
		data = append(data, object)
	}
	writeListEnvelope(w, data)
	return nil
}

// textParam mirrors queryString(...) consumers: only single string values
// count; repeated params become arrays in express and are ignored.
func textParam(query url.Values, name string) string {
	values := query[name]
	if len(values) != 1 {
		return ""
	}
	return strings.TrimSpace(values[0])
}

func textParamOrEmpty(query url.Values, name string) string {
	return textParam(query, name)
}

func (d *Deps) listContainerFiles(w http.ResponseWriter, r *http.Request) error {
	scope := d.requireScope(w, r)
	if scope == nil {
		return nil
	}
	containerID, err := pathContainerID(r)
	if err != nil {
		return err
	}
	purpose := "code_interpreter_output"
	result, err := d.Store.ListFiles(r.Context(), FileListOptions{
		SystemAccountID: scope.SystemAccountID,
		APIKeyID:        scope.APIKeyID,
		Purpose:         &purpose,
		ContainerID:     &containerID,
		Limit:           queryIntegerParam(r.URL.Query(), "limit"),
		Order:           textParam(r.URL.Query(), "order"),
		After:           textParamOrEmpty(r.URL.Query(), "after"),
	})
	if err != nil {
		return err
	}
	data := make([]any, 0, len(result.Items))
	for _, item := range result.Items {
		object, err := newContainerFileObject(item)
		if err != nil {
			return err
		}
		data = append(data, object)
	}
	writeListEnvelope(w, data)
	return nil
}

func (d *Deps) getFile(w http.ResponseWriter, r *http.Request) error {
	if _, ok := d.requireScopeForLookup(w, r); !ok {
		return nil
	}
	record, err := d.findFileForRequest(w, r)
	if err != nil {
		return err
	}
	if record == nil {
		return notFound("文件不存在", "file_not_found")
	}
	object, err := newFileObject(*record)
	if err != nil {
		return err
	}
	writeJSONBody(w, http.StatusOK, object)
	return nil
}

func (d *Deps) getContainerFile(w http.ResponseWriter, r *http.Request) error {
	if _, ok := d.requireScopeForLookup(w, r); !ok {
		return nil
	}
	record, err := d.findContainerFileForRequest(w, r)
	if err != nil {
		return err
	}
	if record == nil {
		return notFound("容器文件不存在", "container_file_not_found")
	}
	object, err := newContainerFileObject(*record)
	if err != nil {
		return err
	}
	writeJSONBody(w, http.StatusOK, object)
	return nil
}

// requireScopeForLookup keeps the 401-before-404 ordering of the Node
// handlers (requireGatewayRuntime runs before the lookup).
func (d *Deps) requireScopeForLookup(w http.ResponseWriter, r *http.Request) (*GatewayScope, bool) {
	scope := d.requireScope(w, r)
	return scope, scope != nil
}

func (d *Deps) findFileForRequest(w http.ResponseWriter, r *http.Request) (*FileRecord, error) {
	scope := d.scope(r)
	fileID, err := pathFileID(r)
	if err != nil {
		return nil, err
	}
	return d.Store.FindFile(r.Context(), fileID, scope.SystemAccountID, scope.APIKeyID)
}

func (d *Deps) findContainerFileForRequest(w http.ResponseWriter, r *http.Request) (*FileRecord, error) {
	containerID, err := pathContainerID(r)
	if err != nil {
		return nil, err
	}
	record, err := d.findFileForRequest(w, r)
	if err != nil {
		return nil, err
	}
	if record == nil || record.Purpose != "code_interpreter_output" || record.ContainerID == nil || *record.ContainerID != containerID {
		return nil, nil
	}
	return record, nil
}

func (d *Deps) downloadFileContent(w http.ResponseWriter, r *http.Request) error {
	return d.downloadContent(w, r, d.findFileForRequest, "文件不存在", "file_not_found")
}

func (d *Deps) downloadContainerFileContent(w http.ResponseWriter, r *http.Request) error {
	return d.downloadContent(w, r, d.findContainerFileForRequest, "容器文件不存在", "container_file_not_found")
}

func (d *Deps) downloadContent(w http.ResponseWriter, r *http.Request, find func(w http.ResponseWriter, r *http.Request) (*FileRecord, error), message, code string) error {
	if _, ok := d.requireScopeForLookup(w, r); !ok {
		return nil
	}
	record, err := find(w, r)
	if err != nil {
		return err
	}
	if record == nil {
		return badRequest(message, code)
	}
	filePath, err := FileObjectPath(d.filesRoot(), record.StorageKey)
	if err != nil {
		return err
	}
	file, err := os.Open(filePath)
	if err != nil {
		return errUnhandled
	}
	defer file.Close()
	mediaType := "application/octet-stream"
	if record.MediaType != nil && *record.MediaType != "" {
		mediaType = *record.MediaType
	}
	w.Header().Set("Content-Type", mediaType)
	w.Header().Set("Content-Length", strconv.FormatInt(record.Bytes, 10))
	w.WriteHeader(http.StatusOK)
	_, _ = io.Copy(w, file)
	return nil
}

func (d *Deps) deleteFile(w http.ResponseWriter, r *http.Request) error {
	scope := d.requireScope(w, r)
	if scope == nil {
		return nil
	}
	fileID, err := pathFileID(r)
	if err != nil {
		return err
	}
	record, err := d.Store.DeleteFile(r.Context(), fileID, scope.SystemAccountID, scope.APIKeyID)
	if err != nil {
		return err
	}
	if record == nil {
		return notFound("文件不存在", "file_not_found")
	}
	_ = RemoveFileObject(d.filesRoot(), record.StorageKey)
	writeJSONBody(w, http.StatusOK, deletedFileEnvelope{ID: record.ID, Object: "file", Deleted: true})
	return nil
}

// uploadedFile mirrors UploadedOpenAICompatibleFile.
type uploadedFile struct {
	FileID     string
	Filename   string
	MediaType  string
	HasMedia   bool
	StorageKey string
	Bytes      int64
	SHA256     string
}

// uploadFile mirrors uploadOpenAICompatibleFile.
func (d *Deps) uploadFile(w http.ResponseWriter, r *http.Request) error {
	scope := d.requireScope(w, r)
	if scope == nil {
		return nil
	}
	contentType := r.Header.Get("Content-Type")
	if !isMultipartContentType(contentType) {
		return badRequest("文件上传请求必须使用 multipart/form-data", "invalid_content_type")
	}
	purpose, file, err := d.readMultipartUpload(r)
	if err != nil {
		return err
	}
	created, err := d.Store.CreateFile(r.Context(), FileCreateInput{
		ID:              file.FileID,
		SystemAccountID: scope.SystemAccountID,
		APIKeyID:        scope.APIKeyID,
		Purpose:         purpose,
		Filename:        file.Filename,
		Bytes:           file.Bytes,
		MediaType:       mediaTypePointer(file),
		StorageKey:      file.StorageKey,
		SHA256:          file.SHA256,
	})
	if err != nil {
		_ = RemoveFileObject(d.filesRoot(), file.StorageKey)
		return err
	}
	object, err := newFileObject(created)
	if err != nil {
		return err
	}
	writeJSONBody(w, http.StatusOK, object)
	return nil
}

func mediaTypePointer(file *uploadedFile) *string {
	if !file.HasMedia {
		return nil
	}
	return &file.MediaType
}

// readMultipartUpload mirrors readOpenAICompatibleMultipartUpload with the
// busboy limits files:1 / fields:8 / fileSize:maxBytes.
func (d *Deps) readMultipartUpload(r *http.Request) (string, *uploadedFile, error) {
	maxBytes := d.maxFileBytes()
	_, params, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
	if err != nil {
		return "", nil, errUnhandled
	}
	reader := multipart.NewReader(r.Body, params["boundary"])
	purpose := ""
	seenFileParts := 0
	fieldsSeen := 0
	var upload *uploadedFile
	var parseError error
	for {
		part, partErr := reader.NextPart()
		if partErr == io.EOF {
			break
		}
		if partErr != nil {
			// Malformed multipart: Node's busboy 'error' rejects the parse
			// promise and the express error handler renders the 500.
			return "", nil, errUnhandled
		}
		if part.FileName() == "" {
			fieldsSeen++
			if fieldsSeen <= 8 && part.FormName() == "purpose" {
				value, readErr := io.ReadAll(part)
				if readErr != nil {
					part.Close()
					return "", nil, errUnhandled
				}
				purpose = strings.TrimSpace(string(value))
			}
			part.Close()
			continue
		}
		seenFileParts++
		if seenFileParts > 1 {
			parseError = badRequest("每次请求只能上传一个文件", "too_many_files")
			part.Close()
			continue
		}
		if part.FormName() != "file" {
			part.Close()
			continue
		}
		upload, partErr = d.writeUploadedFile(part, maxBytes)
		if partErr != nil {
			if upload != nil {
				_ = RemoveFileObject(d.filesRoot(), upload.StorageKey)
			}
			return "", nil, partErr
		}
		part.Close()
	}
	if parseError != nil {
		// Node throws the parseError before awaiting the pending file writes,
		// so an already-written upload stays on disk (orphan) — kept aligned.
		return "", nil, parseError
	}
	if purpose == "" {
		if upload != nil {
			_ = RemoveFileObject(d.filesRoot(), upload.StorageKey)
		}
		return "", nil, badRequest("缺少必填 multipart 字段：purpose", "missing_purpose")
	}
	if upload == nil {
		return "", nil, badRequest("缺少必填 multipart 文件字段：file", "missing_file")
	}
	return purpose, upload, nil
}

// writeUploadedFile mirrors writeUploadedOpenAICompatibleFile: persist under
// the storage root with a sha256 and size accounting, enforcing the upload
// limit (413) and the empty-file rule (400).
func (d *Deps) writeUploadedFile(part *multipart.Part, maxBytes int64) (*uploadedFile, error) {
	fileID := d.Store.generateID("file")
	filename := NormalizedUploadFilename(part.FileName())
	mediaType := NormalizeFileMediaType(part.Header.Get("Content-Type"), filename)
	storageKey := StorageKeyForFile(fileID)
	filePath, err := EnsureFileObjectParent(d.filesRoot(), storageKey)
	if err != nil {
		return nil, err
	}
	target, err := os.Create(filePath)
	if err != nil {
		return nil, err
	}
	hash := sha256.New()
	counter := &countingHashWriter{hash: hash}
	written, copyErr := io.Copy(io.MultiWriter(target, counter), io.LimitReader(part, maxBytes+1))
	closeErr := target.Close()
	if copyErr == nil {
		copyErr = closeErr
	}
	uploaded := &uploadedFile{
		FileID:     fileID,
		Filename:   filename,
		MediaType:  mediaType,
		HasMedia:   mediaType != "",
		StorageKey: storageKey,
		Bytes:      counter.bytes,
		SHA256:     hex.EncodeToString(hash.Sum(nil)),
	}
	if copyErr != nil || written > maxBytes {
		_ = RemoveFileObject(d.filesRoot(), storageKey)
		if written > maxBytes {
			return nil, newRequestError("上传文件过大", 413, "request_too_large", "file_too_large")
		}
		return nil, errUnhandled
	}
	if counter.bytes <= 0 {
		_ = RemoveFileObject(d.filesRoot(), storageKey)
		return nil, badRequest("上传文件为空", "empty_file")
	}
	return uploaded, nil
}

type countingHashWriter struct {
	hash  interface{ Write([]byte) (int, error) }
	bytes int64
}

func (c *countingHashWriter) Write(p []byte) (int, error) {
	n, err := c.hash.Write(p)
	c.bytes += int64(n)
	return n, err
}
